pkg/operator/customroute/custom_route_controller.go

```diff
@@ -0,0 +1,283 @@
+	targetRoute    *routev1.Route
+	consumingUsers []configv1.ConsumingUser
+}
+
+func NewCustomRouteController(
+	componentRouteNamespace string,
+	componentRouteName string,
+	destSecretNamespace string,
+	destSecretName string,
+	targetRoute *routev1.Route,
+	consumingUsers []configv1.ConsumingUser,
+	ingressInformer configinformers.IngressInformer,
+	ingressClient configsetterv1.IngressInterface,
+	routeInformer routeinformer.RouteInformer,
+	routeClient routeclient.RouteInterface,
+	kubeInformersForNamespaces v1helpers.KubeInformersForNamespaces,
+	operatorClient v1helpers.OperatorClient,
+	eventRecorder events.Recorder,
+	resourceSyncer resourcesynccontroller.ResourceSyncer,
+) factory.Controller {
+	controller := &customRouteController{
+		destSecret: types.NamespacedName{
+			Namespace: destSecretNamespace,
+			Name:      destSecretName,
+		},
+		componentRoute: types.NamespacedName{
+			Namespace: componentRouteNamespace,
+			Name:      componentRouteName,
+		},
+		targetRoute:    targetRoute.DeepCopy(),
+		consumingUsers: consumingUsers,
+		ingressLister:  ingressInformer.Lister(),
+		ingressClient:  ingressClient,
+		routeLister:    routeInformer.Lister(),
+		routeClient:    routeClient,
+		secretLister:   kubeInformersForNamespaces.SecretLister(),
+		operatorClient: operatorClient,
+		resourceSyncer: resourceSyncer,
+	}
+
+	return factory.New().
+		WithInformers(
+			ingressInformer.Informer(),
+			routeInformer.Informer(),
+			kubeInformersForNamespaces.InformersFor("openshift-config").Core().V1().Secrets().Informer(),
+			kubeInformersForNamespaces.InformersFor(destSecretNamespace).Core().V1().Secrets().Informer(),
+		).
+		WithSyncDegradedOnError(operatorClient).
+		WithSync(controller.sync).
+		ResyncEvery(time.Minute).
+		ToController("CustomRouteController", eventRecorder.WithComponentSuffix("custom-route-controller"))
+}
+
+func (c *customRouteController) sync(ctx context.Context, syncCtx factory.SyncContext) error {
+	ingressConfig, err := c.ingressLister.Get("cluster")
+	if err != nil {
+		return err
+	}
+
+	ingressConfigCopy := ingressConfig.DeepCopy()
+
+	// configure the expected route
+	expectedRoute, secretName, errors := c.getOAuthRouteAndSecretName(ingressConfigCopy)
+	if errors != nil {
+		// log if there is an issue updating the ingressConfig resource
+		if updateIngressConfigErr := c.updateIngressConfigStatus(ctx, ingressConfigCopy, errors); updateIngressConfigErr != nil {
+			klog.Infof("Error updating ingress with custom route status: %v", err)
+		}
+		return fmt.Errorf("custom route configuration failed verification: %v", errors)
+	}
+
+	// create or modify the existing route
+	if _, _, err = resourceapply.ApplyRoute(ctx, c.routeClient, syncCtx.Recorder(), expectedRoute); err != nil {
+		return err
+	}
+
+	// update ingressConfig status
+	if err = c.updateIngressConfigStatus(ctx, ingressConfigCopy, nil); err != nil {
+		return err
+	}
+
+	// sync the secret
+	return c.syncSecret(secretName)
```

Comment:
looks pretty weird to do work *after* you've set the condition above.

Reply:
Syncing secret mechanism uses `resourcesynccontroller.ResourceSyncer`. The syncer validates if source or destination namespaces and names are set and then syncs Secrets/ConfigMaps asynchronously.

Because of this design, the only way (at least that I know) to know if anything goes wrong is to observe the `Conditions` in the Operator's status. Checking the error here wouldn't buy us anything. 

If you know any better way to do it - please let me know. Perhaps, this piece of code could be improved.